roachtest: scaledata/filesystem_simulator/nodes=6 failed

SHA: https://github.com/cockroachdb/cockroach/commits/57274fc2cd56e5e57957db59acbbcf4f35c118bc

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scaledata/filesystem_simulator/nodes=6 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1229018&tab=buildLog

```
The test failed on master:
	cluster.go:1688,scaledata.go:126,scaledata.go:53,test.go:1237: unexpected node event: 6: dead
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,scaledata.go:126,scaledata.go:53,test.go:1237: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1229018-scaledata-filesystem-simulator-nodes-6 --oneshot --ignore-empty-nodes: exit status 1 7: skipped
		5: 3121
		3: 4515
		2: 3600
		4: 3609
		6: dead
		1: 3074
		Error:  6: dead

```

```
E190409 12:36:03.410257 460380 util/log/crash_reporting.go:204  [n6,merge] a panic has occurred!
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x19f63cd]

goroutine 460380 [running]:
panic(0x2ee7e00, 0x574d230)
	/usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc00d3039e0 sp=0xc00d303950 pc=0x735dfb
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0001c4d80, 0x3a57480, 0xc00999f200)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f fp=0xc00d303a40 sp=0xc00d3039e0 pc=0x10a517f
runtime.call32(0x0, 0x344ba80, 0xc00768d0c0, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc00d303a70 sp=0xc00d303a40 pc=0x7649cb
panic(0x2ee7e00, 0x574d230)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc00d303b00 sp=0xc00d303a70 pc=0x735ce9
runtime.panicmem()
	/usr/local/go/src/runtime/panic.go:82 +0x5e fp=0xc00d303b20 sp=0xc00d303b00 pc=0x734e3e
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:390 +0x182 fp=0xc00d303b70 sp=0xc00d303b20 pc=0x74be22
github.com/cockroachdb/cockroach/pkg/storage.(*replicaItem).setProcessing(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:80
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).addToPurgatoryLocked.func2.1(0xc0000d5ce0, 0x3a57480, 0xc00999f200, 0xc0001c4d80)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:928 +0x1dd fp=0xc00d303ca8 sp=0xc00d303b70 pc=0x19f63cd
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).addToPurgatoryLocked.func2(0x3a57480, 0xc00999f200)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:949 +0x462 fp=0xc00d303f88 sp=0xc00d303ca8 pc=0x19f6b42
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00d7f1b90, 0xc0001c4d80, 0xc00d3bcce0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xe1 fp=0xc00d303fc8 sp=0xc00d303f88 pc=0x10a7301
```

That one's probably on me.

--- a/pkg/storage/queue.go
+++ b/pkg/storage/queue.go
@@ -924,10 +924,11 @@ func (bq *baseQueue) addToPurgatoryLocked(
 					bq.mu.Lock()
 					ranges := make([]roachpb.RangeID, 0, len(bq.mu.purgatory))
 					for rangeID := range bq.mu.purgatory {
-						item := bq.mu.replicas[rangeID]
-						item.setProcessing()
-						ranges = append(ranges, item.value)
-						bq.removeFromPurgatoryLocked(item)
+						if item, ok := bq.mu.replicas[rangeID]; ok {
+							item.setProcessing()
+							ranges = append(ranges, item.value)
+						}
+						bq.removeFromPurgatoryLocked(rangeID)
 					}
 					bq.mu.Unlock()
 
@@ -1019,7 +1020,7 @@ func (bq *baseQueue) removeLocked(item *replicaItem) {
 		item.requeue = false
 	} else {
 		if _, ok := bq.mu.purgatory[item.value]; ok {
-			bq.removeFromPurgatoryLocked(item)
+			bq.removeFromPurgatoryLocked(item.value)
 		} else {
 			bq.removeFromQueueLocked(item)
 		}
@@ -1028,8 +1029,8 @@ func (bq *baseQueue) removeLocked(item *replicaItem) {
 }
 
 // Caller must hold mutex.
-func (bq *baseQueue) removeFromPurgatoryLocked(item *replicaItem) {
-	delete(bq.mu.purgatory, item.value)
+func (bq *baseQueue) removeFromPurgatoryLocked(rangeID roachpb.RangeID) {
+	delete(bq.mu.purgatory, rangeID)
 	bq.purgatory.Update(int64(len(bq.mu.purgatory)))
 }
 

